### PR TITLE
Two bug fixes - disallow "new (int, int)()" and induce a failure if ValueTuple is not a struct.

### PR DESCRIPTION
**Pull request by VSadov**
Fixes: #10891      - Forbid constructing tuples via "new"
Fixes: #11689      - Enforce that ValueTuple types are structs

Note that #11689 requires validation that ValueTuple is a value type only in common cases, That is - to convey that cases where ValueTuple happens to be a Class, Enum, Delegate, etc are clearly not supported.

From the language prospective, tuples are value types composed of their elements. That is visible through the semantics of definite assignment, conversions, nullable lifting, async capturing and so on. This change intentionally introduces emit failures when underlying ValueTuple types are not structs.

This is only C# part. 
VB will need exactly same fixes ported.

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -3181,6 +3181,12 @@ protected BoundExpression BindObjectCreationExpression(ObjectCreationExpressionS
         {
             var type = BindType(node.Type, diagnostics);
 
+            if (node.Type.Kind() == SyntaxKind.TupleType)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Can you please move this constraint into the parser? It is likely to be specified as a syntactic constraint. The syntax `new (x, y)` is reasonably likely to be a work item in the not too distant future (#35). Having this handled in the parser would make the next person's job much easier.

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs [hunk 2]
```diff
@@ -21,7 +21,7 @@ internal abstract class PEModuleBuilder : PEModuleBuilder<CSharpCompilation, Sou
         // TODO: Need to estimate amount of elements for this map and pass that value to the constructor. 
         protected readonly ConcurrentDictionary<Symbol, Cci.IModuleReference> AssemblyOrModuleSymbolToModuleRefMap = new ConcurrentDictionary<Symbol, Cci.IModuleReference>();
         private readonly ConcurrentDictionary<Symbol, object> _genericInstanceMap = new ConcurrentDictionary<Symbol, object>();
-        private readonly ConcurrentSet<ErrorTypeSymbol> _reportedErrorTypesMap = new ConcurrentSet<ErrorTypeSymbol>();
+        private readonly ConcurrentSet<TypeSymbol> _reportedErrorTypesMap = new ConcurrentSet<TypeSymbol>();
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: I think this set should use an identity comparer.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 1 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Will try move it to parser.

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs (hunk 2 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: Will change to identity.
Note that this is an optimization for the number of errors spitted out in extremely rare and badly broken cases. I do not think it makes a lot of difference either way.

**Comment by VSadov** [comment]
@dotnet-bot test perf_correctness_prtest please
@dotnet-bot test windows_release_unit64_prtest please

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 3]
```diff
@@ -10676,7 +10676,13 @@ private ExpressionSyntax ParseArrayOrObjectCreationExpression()
                         SyntaxFactory.MissingToken(SyntaxKind.CloseParenToken));
                 }
 
-                return _syntaxFactory.ObjectCreationExpression(@new, type, argumentList, initializer);
+                var objectCreation = _syntaxFactory.ObjectCreationExpression(@new, type, argumentList, initializer);
+                if (type.Kind == SyntaxKind.TupleType)
```
Comment (gafter) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Might be nicer to put the error on the tuple type.

File: src/Compilers/CSharp/Test/Syntax/Parsing/StatementParsingTests.cs [hunk 4]
```diff
@@ -2597,7 +2597,7 @@ class C1
 {
     static void Test(int arg1, (byte, byte) arg2)
     {
-        (int, int) t1 = new(int, int)();
+        (int, int)? t1 = new(int, int)?();
```
Comment (gafter) [inline src/Compilers/CSharp/Test/Syntax/Parsing/StatementParsingTests.cs]: Can you please modify the test to assert what the diagnostics are? This syntax should also not work (i.e. a nullable tuple type syntax with `new` should also be a syntax error).

Comment (VSadov) [inline src/Compilers/CSharp/Test/Syntax/Parsing/StatementParsingTests.cs]: What about   ```new (int, int)[0] t1```   or  ```new (int, int)?[0] t1``` ?

Comment (VSadov) [inline src/Compilers/CSharp/Test/Syntax/Parsing/StatementParsingTests.cs]: I think we need more design discussions for additional restrictions.

So far we agreed to not allow ```new``` directly with tuple types. 
This change implements that decision.
It started with not being able to do that anyways for tuple types with 7+ elements. With nullables or arrays there are no problems on their own, so we need to formalize another motivation.

We should discuss addition restrictions more comprehensively, just to be sure there is a common rational reason for all the cases and that we covered them all.

Comment (VSadov) [inline src/Compilers/CSharp/Test/Syntax/Parsing/StatementParsingTests.cs]: Just to mention another case: 
``` new (int, int)?(<value>)```   is also legal

```Nullable<T>```  has a ctor that takes T, so as long as ```<value>``` converts to T, the code above has meaning.
So will need to decide whether to allow it or not.

Comment (VSadov) [inline src/Compilers/CSharp/Test/Syntax/Parsing/StatementParsingTests.cs]: Per current rules only a "new" applied immediately to a tuple type is not permitted. So this is valid. For now.

I am afraid generalizing the rule to a wider range of syntax is not so straightforward and unambiguous. I expect a range of opinions and valid reasons why we need to go one way or another from here, so I am not brave enough to guess :-)

If we want to cover a wider range of syntax, we need to tweak the design and implement it in another change.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 5]
```diff
@@ -3181,6 +3181,12 @@ protected BoundExpression BindObjectCreationExpression(ObjectCreationExpressionS
...
+            {
+                Debug.Assert(node.HasErrors, "new <tuple type> should be a syntax error");
+                return BadExpression(node, LookupResultKind.NotCreatable);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Should `type` be used for the type?

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs [hunk 6]
```diff
@@ -21,7 +21,7 @@ internal abstract class PEModuleBuilder : PEModuleBuilder<CSharpCompilation, Sou
         // TODO: Need to estimate amount of elements for this map and pass that value to the constructor. 
         protected readonly ConcurrentDictionary<Symbol, Cci.IModuleReference> AssemblyOrModuleSymbolToModuleRefMap = new ConcurrentDictionary<Symbol, Cci.IModuleReference>();
         private readonly ConcurrentDictionary<Symbol, object> _genericInstanceMap = new ConcurrentDictionary<Symbol, object>();
-        private readonly ConcurrentSet<ErrorTypeSymbol> _reportedErrorTypesMap = new ConcurrentSet<ErrorTypeSymbol>();
+        private readonly ConcurrentSet<TypeSymbol> _reportedErrorTypesMap = new ConcurrentSet<TypeSymbol>(ReferenceEqualityComparer.Instance);
```
Comment (cston) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: Why use reference equality rather than `Equals`? Won't we report more errors as a result?

Comment (gafter) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: These are always original types or error types, which are equal by identity.

File: src/Compilers/CSharp/Test/Syntax/Parsing/StatementParsingTests.cs (hunk 4 above)
Comment (gafter) [inline src/Compilers/CSharp/Test/Syntax/Parsing/StatementParsingTests.cs]: Agreed, but please do check the specific diagnostics here.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 5 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Considering that it might parse into something else in the future, I do not want to bind the type.

File: src/Compilers/CSharp/Test/Syntax/Parsing/StatementParsingTests.cs (hunk 4 above)
Comment (VSadov) [inline src/Compilers/CSharp/Test/Syntax/Parsing/StatementParsingTests.cs]: This test does not result in any diagnostics. (only valid cases are here) and there is validation that we do not have diagnostics.

Once some of this cases become errors, the test will catch that.

Comment (gafter) [inline src/Compilers/CSharp/Test/Syntax/Parsing/StatementParsingTests.cs]: Oh, I mistakenly thought it was asserting there were errors.

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs (hunk 6 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: I did not realize that, just followed Neal's advise, but yes, they are definitions.

---
In reply to: [81423689](https://github.com/dotnet/roslyn/pull/14161#discussion_r81423689) [](ancestors = 81423689)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 3 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: I tried to figure what is at fault here - the "new" or the tuple literal. And ended up deciding to flag the whole thing since it is the combination that is invalid.

---
In reply to: [81399515](https://github.com/dotnet/roslyn/pull/14161#discussion_r81399515) [](ancestors = 81399515)

File: src/Compilers/CSharp/Test/Syntax/Parsing/StatementParsingTests.cs (hunk 4 above)
Comment (VSadov) [inline src/Compilers/CSharp/Test/Syntax/Parsing/StatementParsingTests.cs]: Wont Fix for this PR. 
We need a better understanding of desired design for this change

---
In reply to: [81405544](https://github.com/dotnet/roslyn/pull/14161#discussion_r81405544) [](ancestors = 81405544)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 5 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Binding the type might result in more diagnostics, or opposite - something that uses the expression type may actually work. And that will change once we change parsing. It seems to be better if we do not bind.

Comment (cston) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Aren't we already binding the type at line 3182?

Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Ah. I see.  I meant to move this to after the check. Somehow type binding is still there. Will fix.

---
In reply to: [81443846](https://github.com/dotnet/roslyn/pull/14161#discussion_r81443846) [](ancestors = 81443846)

Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Thanks for noticing this. Fixed.

---
In reply to: [81451427](https://github.com/dotnet/roslyn/pull/14161#discussion_r81451427) [](ancestors = 81451427,81443846)

**Comment by VSadov** [comment]
@dotnet/roslyn-compiler  PING.

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs [hunk 7]
```diff
@@ -790,6 +790,19 @@ protected virtual Cci.IModuleReference TranslateModule(ModuleSymbol module, Diag
...
+                // but if it does happen we should make it a failure.
+                var declaredBase = namedTypeSymbol.BaseTypeNoUseSiteDiagnostics;
+                if (declaredBase.SpecialType != SpecialType.System_ValueType && !declaredBase.IsErrorType())
+                {
+                    // Try to decrease noise by not complaining about the same type over and over again.
+                    if (_reportedErrorTypesMap.Add(namedTypeSymbol))
+                    {
+                        diagnostics.Add(new CSDiagnostic(new CSDiagnosticInfo(ErrorCode.ERR_PredefinedValueTupleTypeMustBeStruct, namedTypeSymbol.MetadataName), syntaxNodeOpt == null ? NoLocation.Singleton : syntaxNodeOpt.Location));
+                    }
+                }
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: Consider extracting this block to a method.
Also, there may be a simpler way to report the diagnostic.

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs [hunk 8]
```diff
@@ -790,6 +790,19 @@ protected virtual Cci.IModuleReference TranslateModule(ModuleSymbol module, Diag
             {
                 Debug.Assert(!needDeclaration);
                 namedTypeSymbol = namedTypeSymbol.TupleUnderlyingType;
+
+                // check that underlying type of a ValueTuple is indeed a value type (or error)
+                // this should never happen, in theory,
+                // but if it does happen we should make it a failure.
+                var declaredBase = namedTypeSymbol.BaseTypeNoUseSiteDiagnostics;
+                if (declaredBase.SpecialType != SpecialType.System_ValueType && !declaredBase.IsErrorType())
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: Is there a reason not to simply check `namedTypeSymbol.IsValueType`? Note that `TupleTypeSymbol.IsValueType` already defaults to true if the underlying type is an error type.

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 9]
```diff
@@ -4930,6 +4930,12 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="WRN_TupleLiteralNameMismatch_Title" xml:space="preserve">
     <value>The tuple element name is ignored because a different name is specified by the assignment target.</value>
   </data>
+  <data name="ERR_PredefinedValueTupleTypeMustBeStruct" xml:space="preserve">
+    <value>Predefined type '{0}' must be a struct.</value>
+  </data>
+  <data name="ERR_NewWithTupleTypeSyntax" xml:space="preserve">
+    <value>"new" cannot be used with tuple type '{0}'. Use a tuple literal expression instead.</value>
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: Resource strings typically using single quotes rather than double quotes for keywords: `'new'` rather than `"new"`

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs [hunk 10]
```diff
@@ -10676,7 +10676,13 @@ private ExpressionSyntax ParseArrayOrObjectCreationExpression()
...
+                {
+                    objectCreation = this.AddError(objectCreation, objectCreation, ErrorCode.ERR_NewWithTupleTypeSyntax, type.ToString());
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: If we have add this error during parsing, why do we also need special handling in `BindObjectCreationExpression`?

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs (hunk 8 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: namedTypeSymbol is not necessarily a tuple and many types would default to be classes or whatever.
We do not want Enums, which are value types, so checking for a base ValueType is actually simpler and more precise.

---
In reply to: [81642661](https://github.com/dotnet/roslyn/pull/14161#discussion_r81642661) [](ancestors = 81642661)

File: src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs [hunk 11]
```diff
@@ -1097,10 +1081,9 @@ public override TypeKind TypeKind
         {
             get
             {
-                // only classes and structs can have instance fields as tuple requires.
-                // we need to have some support for classes, but most common case will be struct
-                // in broken scenarios (ErrorType, Enum, Delegate, whatever..) we will just default to struct.
-                return _underlyingType.TypeKind == TypeKind.Class ? TypeKind.Class : TypeKind.Struct;
+                // From the language prospective tuple is a value type
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs]: >prospective [](start = 37, length = 11)

s/prospective/perspective/

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 10 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: By default we do attempt to bind syntax even if it has errors. We should not in his case. It is intentional that it means nothing now because we expect it to mean a different thing in the future.

---
In reply to: [81643102](https://github.com/dotnet/roslyn/pull/14161#discussion_r81643102) [](ancestors = 81643102)

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs [hunk 12]
```diff
@@ -261,8 +261,8 @@ .maxstack  6
   IL_002f:  ldstr      ""y""
   IL_0034:  call       ""(int A, int B) Microsoft.VisualStudio.Debugger.Clr.IntrinsicMethods.GetVariableAddress<(int A, int B)>(string)""
   IL_0039:  ldloc.0
-  IL_003a:  stind.ref
-  IL_003b:  ret
+  IL_003a:  stobj      ""System.ValueTuple<int, int>""
```
Comment (jcouv) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs]: I didn't understand why this IL changed. What caused this?

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 9 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/CSharpResources.resx]: fixed

---
In reply to: [81642694](https://github.com/dotnet/roslyn/pull/14161#discussion_r81642694) [](ancestors = 81642694)

File: src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs (hunk 11 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Symbols/Tuples/TupleTypeSymbol.cs]: fixed

---
In reply to: [81643451](https://github.com/dotnet/roslyn/pull/14161#discussion_r81643451) [](ancestors = 81643451)

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs (hunk 12 above)
Comment (VSadov) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs]: Tuple type was not a struct here

---
In reply to: [81644123](https://github.com/dotnet/roslyn/pull/14161#discussion_r81644123) [](ancestors = 81644123)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs [hunk 13]
```diff
@@ -18792,6 +18665,229 @@ static void Main()
...
+
+" + tupleattributes_cs;
+
+            var comp = CreateCompilationWithMscorlib(source, assemblyName: "ValueTupleNotStruct", options: TestOptions.DebugExe);
+
+            comp.VerifyEmitDiagnostics(
+                // (9,5): error CS8180: Predefined type 'ValueTuple`2' must be a struct.
+                //     {
+                Diagnostic(ErrorCode.ERR_PredefinedValueTupleTypeMustBeStruct, @"{
+        return (1, 1);
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: The area for squiggles seems larger than necessary. Why isn't the error just located on the tuple literal or the method return type?

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs (hunk 12 above)
Comment (VSadov) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs]: It is an error type because testcase does not reference System facade where ValueType lives. As a result ValueTuple was considered a class.
This code would not be emittable in regular compile scenario. It would have diagnostics about missing references.

---
In reply to: [81644864](https://github.com/dotnet/roslyn/pull/14161#discussion_r81644864) [](ancestors = 81644864,81644123)

Comment (VSadov) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs]: It is really a result of the changes that makes tuples value types regardless whether backing types are errors or not.

---
In reply to: [81645389](https://github.com/dotnet/roslyn/pull/14161#discussion_r81645389) [](ancestors = 81645389,81644864,81644123)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs (hunk 13 above)
Comment (VSadov) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs]: this is what is passed to that method as a relevant syntax reference from the callers. 
I think this is sufficient. This is going to be very rare diagnostics. It does not need to be very user friendly.

---
In reply to: [81644930](https://github.com/dotnet/roslyn/pull/14161#discussion_r81644930) [](ancestors = 81644930)

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs (hunk 8 above)
Comment (cston) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: `declaredBase` will be `null` for `interface ValueTuple<T1, T2>`.

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs (hunk 8 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: Thanks!! Fixed and added a test.

---
In reply to: [81646014](https://github.com/dotnet/roslyn/pull/14161#discussion_r81646014) [](ancestors = 81646014)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 10 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: >type.ToString() [](start = 121, length = 15)

Do we do this anywhere else? How does this look with line breaks within the type syntax?

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs (hunk 6 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: In general, error types are not equal by identity. Please revert the change.

---
In reply to: [81434867](https://github.com/dotnet/roslyn/pull/14161#discussion_r81434867) [](ancestors = 81434867,81423689)

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs [hunk 14]
```diff
@@ -890,6 +891,23 @@ protected virtual Cci.IModuleReference TranslateModule(ModuleSymbol module, Diag
...
+            // this should never happen, in theory,
+            // but if it does happen we should make it a failure.
+            // NOTE: declaredBase could be null for interfaces
+            var declaredBase = namedTypeSymbol.BaseTypeNoUseSiteDiagnostics;
+            if (declaredBase?.SpecialType != SpecialType.System_ValueType && declaredBase?.IsErrorType() != false)
+            {
+                // Try to decrease noise by not complaining about the same type over and over again.
+                if (_reportedErrorTypesMap.Add(namedTypeSymbol))
+                {
+                    diagnostics.Add(new CSDiagnostic(new CSDiagnosticInfo(ErrorCode.ERR_PredefinedValueTupleTypeMustBeStruct, namedTypeSymbol.MetadataName), syntaxNodeOpt == null ? NoLocation.Singleton : syntaxNodeOpt.Location));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: >diagnostics.Add(new CSDiagnostic(new CSDiagnosticInfo(ErrorCode.ERR_PredefinedValueTupleTypeMustBeStruct, namedTypeSymbol.MetadataName), syntaxNodeOpt == null ? NoLocation.Singleton : syntaxNodeOpt.Location)); [](start = 20, length = 209)

Do we have a test for a nested case.

Comment (VSadov) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: nested as in ... example?

---
In reply to: [81654602](https://github.com/dotnet/roslyn/pull/14161#discussion_r81654602) [](ancestors = 81654602)

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs (hunk 6 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: Will revert

---
In reply to: [81654084](https://github.com/dotnet/roslyn/pull/14161#discussion_r81654084) [](ancestors = 81654084,81434867,81423689)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 10 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: It is more common that we do "currentToken.Text" which also can be multiline. 
I think the error will not be nice looking.  Multiline errors are not nice looking in general. 

I do not think there is a lot of choice here other than not specify the type in the error at all.

---
In reply to: [81653476](https://github.com/dotnet/roslyn/pull/14161#discussion_r81653476) [](ancestors = 81653476)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 15]
```diff
@@ -3179,6 +3179,12 @@ private static bool IsNegativeConstantForArraySize(BoundExpression expression)
 
         protected BoundExpression BindObjectCreationExpression(ObjectCreationExpressionSyntax node, DiagnosticBag diagnostics)
         {
+            if (node.Type.Kind() == SyntaxKind.TupleType)
+            {
+                Debug.Assert(node.HasErrors, "new <tuple type> should be a syntax error");
+                return BadExpression(node, LookupResultKind.NotCreatable);
```
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs]: Agreed. If we happened to parse it (not make a junk trivia), we may as well bind it. 
Perhaps should put "HasError" flah on the bound node though. To make sure it does not go further.

---
In reply to: [81652702](https://github.com/dotnet/roslyn/pull/14161#discussion_r81652702) [](ancestors = 81652702)

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs (hunk 12 above)
Comment (AlekseyTs) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs]: Could you clarify why this isn't an error in this case?

---
In reply to: [81645590](https://github.com/dotnet/roslyn/pull/14161#discussion_r81645590) [](ancestors = 81645590,81645389,81644864,81644123)

**Comment by AlekseyTs** [comment]
Do we have a test verifying that one can still invoke constructors by using type name?
```
new ValueTuple<int, long>(1, 2)
```

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs (hunk 14 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: A tuple of cardinality >= 8 is used and VT`8 is not a structure, or "short" VT is not a structure, or both are not structures.

**Comment by VSadov** [comment]
Yes, we have those. UnifyUnderlyingWithTuple_08 has a bunch.

---
In reply to: [251253047](https://github.com/dotnet/roslyn/pull/14161#issuecomment-251253047) [](ancestors = 251253047)

File: src/Compilers/CSharp/Portable/Parser/LanguageParser.cs (hunk 10 above)
Comment (VSadov) [inline src/Compilers/CSharp/Portable/Parser/LanguageParser.cs]: Error gets truncated. It does look ugly. 
I will put squigglies only for the type and not mention the actual type i the error

---
In reply to: [81658408](https://github.com/dotnet/roslyn/pull/14161#discussion_r81658408) [](ancestors = 81658408,81653476)

**Comment by VSadov** [comment]
@dotnet-bot test windows_debug_unit32_prtest  please

File: src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs [hunk 16]
```diff
@@ -890,6 +891,23 @@ protected virtual Cci.IModuleReference TranslateModule(ModuleSymbol module, Diag
...
         }
 
+        private void CheckTupleUnderlying(NamedTypeSymbol namedTypeSymbol, SyntaxNode syntaxNodeOpt, DiagnosticBag diagnostics)
+        {
+            // check that underlying type of a ValueTuple is indeed a value type (or error)
+            // this should never happen, in theory,
+            // but if it does happen we should make it a failure.
+            // NOTE: declaredBase could be null for interfaces
+            var declaredBase = namedTypeSymbol.BaseTypeNoUseSiteDiagnostics;
+            if (declaredBase?.SpecialType != SpecialType.System_ValueType && declaredBase?.IsErrorType() != true)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: > != true [](start = 104, length = 8)

Is there a test that backs this change?

Comment (VSadov) [inline src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs]: It covers the case where ValueTuple is an interface. (and it was indeed causing a crash)

---
In reply to: [81813233](https://github.com/dotnet/roslyn/pull/14161#discussion_r81813233) [](ancestors = 81813233,81812665)

File: src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs (hunk 12 above)
Comment (cston) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs]: Logged #14267 to report a compile error here.

Comment (VSadov) [inline src/ExpressionEvaluator/CSharp/Test/ExpressionCompiler/TupleTests.cs]: Thanks! We should either recognize missing ValueType to be a special type (that would make derived types structs) or report an error. Neither is happening. 

---
In reply to: [81822455](https://github.com/dotnet/roslyn/pull/14161#discussion_r81822455) [](ancestors = 81822455)
